# Merging Subcounty Population Data from the Uganda Census with Antiretroviral Service Data from October-December, 2024

In [1]:
import pandas as pd

In [2]:
northern = pd.read_csv('Central region_2014 - 1.Population and HH Composition.csv')

In [3]:
northern.head()

District   County Subcounty    Parish Population Size by Parish  \
0        NaN      NaN       NaN       NaN                      Male   
1        NaN      NaN       NaN       NaN                       NaN   
2  Kalangala  Bujumba   Bujumba   Bujumba                      1096   
3  Kalangala  Bujumba   Bujumba   Bunyama                      1428   
4  Kalangala  Bujumba   Bujumba  Bwendero                      2214   

  Unnamed: 5  Unnamed: 6 Unnamed: 7 Headship of Households by Sex Unnamed: 9  \
0     Female  Sex Ratio       Total                   Male Headed        NaN   
1        NaN         NaN        NaN                        Number    Percent   
2        874       125.4       1970                           479       80.9   
3        944       151.3       2372                           806       83.8   
4       1627       136.1       3841                          1284         81   

   ...         Unnamed: 54 Unnamed: 55         Unnamed: 56 Unnamed: 57  \
0  ...  10 Years and Above         NaN  15 Years and Above         NaN   
1  ...              Number     Percent              Number     Percent   
2  ...                1439        73.1                1233        62.7   
3  ...                1679        73.6                1586        69.5   
4  ...                2807        74.7                2667          71   

          Unnamed: 58 Unnamed: 59         Unnamed: 60 Unnamed: 61  \
0  18 Years and above         NaN  20 Years and above         NaN   
1              Number     Percent              Number     Percent   
2                1120        56.9                1038        52.7   
3                1526        66.9                1460          64   
4                2562        68.2                2419        64.4   

          Unnamed: 62 Unnamed: 63  
0  65 Years and Above         NaN  
1              Number     Percent  
2                  63         3.2  
3                  21         0.9  
4                  31         0.8  

[5 rows x 64 columns]

In [4]:
eastern = pd.read_csv('Eastern region_2014 - 1.Pop. & HH Composition.csv')

In [5]:
eastern.head()

District               County         Subcounty      Parish  \
0      NaN                  NaN               NaN         NaN   
1      NaN                  NaN               NaN         NaN   
2   Bugiri  Bugiri Municipality  Eastern Division  Naluwerere   
3   Bugiri  Bugiri Municipality  Eastern Division       Nkusi   
4   Bugiri  Bugiri Municipality  Western Division       Bwole   

  Population Size by Parish Unnamed: 5  Unnamed: 6 Unnamed: 7  \
0                      Male     Female  Sex Ratio       Total   
1                       NaN        NaN         NaN        NaN   
2                      2798       3080        90.8       5878   
3                      3682       4264        86.4       7946   
4                      3919       4451          88       8370   

  Headship of Households by Sex Unnamed: 9  ...         Unnamed: 54  \
0                   Male Headed        NaN  ...  10 Years and Above   
1                        Number    Percent  ...              Number   
2                          1041       74.8  ...                3842   
3                          1523       74.3  ...                5365   
4                          1535       74.5  ...                5563   

  Unnamed: 55         Unnamed: 56 Unnamed: 57         Unnamed: 58 Unnamed: 59  \
0         NaN  15 Years and Above         NaN  18 Years and above         NaN   
1     Percent              Number     Percent              Number     Percent   
2        65.5                2981        50.8                2555        43.6   
3        69.3                4346        56.2                3752        48.5   
4          68                4480        54.8                3868        47.3   

          Unnamed: 60 Unnamed: 61         Unnamed: 62 Unnamed: 63  
0  20 Years and above         NaN  65 Years and Above         NaN  
1              Number     Percent              Number     Percent  
2                2324        39.6                 108         1.8  
3                3329          43                  76           1  
4                3424        41.8                 115         1.4  

[5 rows x 64 columns]

In [6]:
western = pd.read_csv('Western region_2014 - 1.Population and HH Composition.csv')

In [7]:
western.head()

District      County Subcounty     Parish Population Size by Parish  \
0         NaN         NaN       NaN        NaN                      Male   
1         NaN         NaN       NaN        NaN                       NaN   
2  Bundibugyo  Bughendera   Bukonzo    Buhundu                      1319   
3  Bundibugyo  Bughendera   Bukonzo  Bukangama                      1143   
4  Bundibugyo  Bughendera   Bukonzo    Bunguha                       608   

  Unnamed: 5  Unnamed: 6 Unnamed: 7 Headship of Households by Sex Unnamed: 9  \
0     Female  Sex Ratio       Total                   Male Headed        NaN   
1        NaN         NaN        NaN                        Number    Percent   
2       1342        98.3       2661                           380       84.8   
3       1160        98.5       2303                           341       83.8   
4        629        96.7       1237                           176       84.2   

   ...         Unnamed: 54 Unnamed: 55         Unnamed: 56 Unnamed: 57  \
0  ...  10 Years and Above         NaN  15 Years and Above         NaN   
1  ...              Number     Percent              Number     Percent   
2  ...                1680        63.1                1274        47.9   
3  ...                1464        63.6                1094        47.5   
4  ...                 839        67.8                 653        52.8   

          Unnamed: 58 Unnamed: 59         Unnamed: 60 Unnamed: 61  \
0  18 Years and above         NaN  20 Years and above         NaN   
1              Number     Percent              Number     Percent   
2                1123        42.2                 997        37.5   
3                 939        40.8                 848        36.8   
4                 532          43                 439        35.5   

          Unnamed: 62 Unnamed: 63  
0  65 Years and Above         NaN  
1              Number     Percent  
2                  55         2.1  
3                  61         2.6  
4                  28         2.3  

[5 rows x 64 columns]

In [8]:
central = pd.read_csv('Central region_2014 - 1.Population and HH Composition.csv')

In [9]:
central.head()

District   County Subcounty    Parish Population Size by Parish  \
0        NaN      NaN       NaN       NaN                      Male   
1        NaN      NaN       NaN       NaN                       NaN   
2  Kalangala  Bujumba   Bujumba   Bujumba                      1096   
3  Kalangala  Bujumba   Bujumba   Bunyama                      1428   
4  Kalangala  Bujumba   Bujumba  Bwendero                      2214   

  Unnamed: 5  Unnamed: 6 Unnamed: 7 Headship of Households by Sex Unnamed: 9  \
0     Female  Sex Ratio       Total                   Male Headed        NaN   
1        NaN         NaN        NaN                        Number    Percent   
2        874       125.4       1970                           479       80.9   
3        944       151.3       2372                           806       83.8   
4       1627       136.1       3841                          1284         81   

   ...         Unnamed: 54 Unnamed: 55         Unnamed: 56 Unnamed: 57  \
0  ...  10 Years and Above         NaN  15 Years and Above         NaN   
1  ...              Number     Percent              Number     Percent   
2  ...                1439        73.1                1233        62.7   
3  ...                1679        73.6                1586        69.5   
4  ...                2807        74.7                2667          71   

          Unnamed: 58 Unnamed: 59         Unnamed: 60 Unnamed: 61  \
0  18 Years and above         NaN  20 Years and above         NaN   
1              Number     Percent              Number     Percent   
2                1120        56.9                1038        52.7   
3                1526        66.9                1460          64   
4                2562        68.2                2419        64.4   

          Unnamed: 62 Unnamed: 63  
0  65 Years and Above         NaN  
1              Number     Percent  
2                  63         3.2  
3                  21         0.9  
4                  31         0.8  

[5 rows x 64 columns]

In [10]:
# First, ensure only the required columns are selected from each dataframe
columns_to_keep = ['District', 'County', 'Subcounty', 'Parish', 'Unnamed: 7']

northern = northern[columns_to_keep]
eastern = eastern[columns_to_keep]
western = western[columns_to_keep]
central = central[columns_to_keep]

# Concatenate the dataframes (since they likely have distinct 'Subcounty' values)
merged_df = pd.concat([northern, eastern, western, central], ignore_index=True)

# If you actually want to *merge* on 'Subcounty' (i.e. combine rows with matching 'Subcounty'), use merge instead:
# merged_df = northern.merge(eastern, on='Subcounty', how='outer')[columns_to_keep]  # Only if 'Subcounty' is shared

# Display or process the result
print(merged_df.head())

    District   County Subcounty    Parish Unnamed: 7
0        NaN      NaN       NaN       NaN      Total
1        NaN      NaN       NaN       NaN        NaN
2  Kalangala  Bujumba   Bujumba   Bujumba       1970
3  Kalangala  Bujumba   Bujumba   Bunyama       2372
4  Kalangala  Bujumba   Bujumba  Bwendero       3841


In [11]:
merged_df

District   County Subcounty      Parish Unnamed: 7
0           NaN      NaN       NaN         NaN      Total
1           NaN      NaN       NaN         NaN        NaN
2     Kalangala  Bujumba   Bujumba     Bujumba       1970
3     Kalangala  Bujumba   Bujumba     Bunyama       2372
4     Kalangala  Bujumba   Bujumba    Bwendero       3841
...         ...      ...       ...         ...        ...
7441    Kyotera  Kyotera  Nabigasa  Bethelehem       5102
7442    Kyotera  Kyotera  Nabigasa     Kijejja       1761
7443    Kyotera  Kyotera  Nabigasa   Kyassimbi       3713
7444    Kyotera  Kyotera  Nabigasa    Nabigasa       5448
7445    Kyotera  Kyotera  Nabigasa   Nakatoogo       4634

[7446 rows x 5 columns]

In [12]:
merged_df = merged_df.drop(1)

In [13]:
merged_df = merged_df.drop(0).reset_index()

In [14]:
merged_df = merged_df.reset_index()

In [15]:
merged_df['Population Size by Parish'] = merged_df['Unnamed: 7']

In [16]:
merged_df

level_0  index   District   County               Subcounty       Parish  \
0           0      2  Kalangala  Bujumba                 Bujumba      Bujumba   
1           1      3  Kalangala  Bujumba                 Bujumba      Bunyama   
2           2      4  Kalangala  Bujumba                 Bujumba     Bwendero   
3           3      5  Kalangala  Bujumba                 Bujumba     Mulabana   
4           4      6  Kalangala  Bujumba  Kalangala Town Council  Kalangala A   
...       ...    ...        ...      ...                     ...          ...   
7439     7439   7441    Kyotera  Kyotera                Nabigasa   Bethelehem   
7440     7440   7442    Kyotera  Kyotera                Nabigasa      Kijejja   
7441     7441   7443    Kyotera  Kyotera                Nabigasa    Kyassimbi   
7442     7442   7444    Kyotera  Kyotera                Nabigasa     Nabigasa   
7443     7443   7445    Kyotera  Kyotera                Nabigasa    Nakatoogo   

     Unnamed: 7 Population Size by Parish  
0          1970                      1970  
1          2372                      2372  
2          3841                      3841  
3          3373                      3373  
4          1270                      1270  
...         ...                       ...  
7439       5102                      5102  
7440       1761                      1761  
7441       3713                      3713  
7442       5448                      5448  
7443       4634                      4634  

[7444 rows x 8 columns]

In [17]:
merged_df = merged_df.drop(['index', 'Unnamed: 7'], axis=1)

In [18]:
merged_df['Population Size by Parish'] = pd.to_numeric(merged_df['Population Size by Parish'], errors='coerce')

merged_df['Population Size by Parish'] = merged_df['Population Size by Parish'].fillna(0).astype(int)

In [19]:
merged_df = merged_df.groupby(['District', 'County', 'Subcounty'])['Population Size by Parish'].sum().reset_index()

In [20]:
merged_df

District                           County            Subcounty  \
0      Amuria                           Amuria            Abarilela   
1      Amuria                           Amuria              Akeriau   
2      Amuria                           Amuria  Amuria Town Council   
3      Amuria                           Amuria              Apeduru   
4      Amuria                           Amuria               Asamuk   
...       ...                              ...                  ...   
1120   Wakiso  Makindye Ssabagabo Municipality      Ndejje Division   
1121   Wakiso             Nansana Municipality    Busukuma Division   
1122   Wakiso             Nansana Municipality       Gombe Division   
1123   Wakiso             Nansana Municipality     Nabweru Division   
1124   Wakiso             Nansana Municipality     Nansana Division   

      Population Size by Parish  
0                         24290  
1                         14227  
2                          7262  
3                         14718  
4                         21385  
...                         ...  
1120                     186982  
1121                      75460  
1122                     153278  
1123                     212530  
1124                     288980  

[1125 rows x 4 columns]

In [21]:
merged_df['Subcounty Population'] = merged_df['Population Size by Parish']

In [22]:
merged_df = merged_df.drop('Population Size by Parish', axis=1)

In [23]:
merged_df

District                           County            Subcounty  \
0      Amuria                           Amuria            Abarilela   
1      Amuria                           Amuria              Akeriau   
2      Amuria                           Amuria  Amuria Town Council   
3      Amuria                           Amuria              Apeduru   
4      Amuria                           Amuria               Asamuk   
...       ...                              ...                  ...   
1120   Wakiso  Makindye Ssabagabo Municipality      Ndejje Division   
1121   Wakiso             Nansana Municipality    Busukuma Division   
1122   Wakiso             Nansana Municipality       Gombe Division   
1123   Wakiso             Nansana Municipality     Nabweru Division   
1124   Wakiso             Nansana Municipality     Nansana Division   

      Subcounty Population  
0                    24290  
1                    14227  
2                     7262  
3                    14718  
4                    21385  
...                    ...  
1120                186982  
1121                 75460  
1122                153278  
1123                212530  
1124                288980  

[1125 rows x 4 columns]

In [24]:
merged_df.to_csv('uganda_pop_2014.csv', index=False)

In [25]:
merged_df[merged_df['County'].str.contains("Busia", na=False)]

District              County         Subcounty  Subcounty Population
170    Busia  Busia Municipality  Eastern Division                 31935
171    Busia  Busia Municipality  Western Division                 22863

In [26]:
art = pd.read_csv('uganda_art_revised.csv')

In [27]:
art.head()

Statistical Region UAIS 2011_ Region  PIRS UnitID     PIRS District   \
0    Northern Region        North East  W3hXYmtaQ4Q    Kotido District   
1    Northern Region      Mid Northern  auzuV39xOTU     Pader District   
2     Eastern Region        North East  J2JiemiIQvD   Katakwi District   
3     Eastern Region        North East  aVQCbe80pl7    Serere District   
4    Northern Region      Mid Northern  yDPmjwoOIxi  Alebtong District   

         PIRS Subcounty                     PIRS Service outlet  \
0  Nakapelimo Subcounty           405 Brigade Health Centre III   
1       Pader Subcounty  5th Military Division Health Centre IV   
2        Usuk Subcounty                  Aakum Health Centre II   
3      Labori Subcounty               Aarapoo Health Centre III   
4       Abako Subcounty                 Abako Health Centre III   

  PIRS Mechanism name  pirs Mechanism ID PIRS US Agency     DATIM ID  ...  \
0        URC DOD/UPDF           160698.0            DOD  OELkOLxvZxB  ...   
1        URC DOD/UPDF           160698.0            DOD  OELkOLxvZxB  ...   
2  AIC Soroti  Region            87041.0            CDC  l86BIz6pDzr  ...   
3  AIC Soroti  Region            87041.0            CDC  JXTSTKGSHxq  ...   
4   JCRC Kigezi Lango            81997.0          USAID  vitFG8P7w6d  ...   

  5 or more months of ARVs dispensed to patient by Unknown Age, Male  \
0                                                NaN                   
1                                                NaN                   
2                                                NaN                   
3                                                NaN                   
4                                                NaN                   

  6 or more months of ARVs dispensed to patient by <15yrs, Female  \
0                                                  0                
1                                                  1                
2                                                  0                
3                                                 56                
4                                                 11                

  6 or more months of ARVs dispensed to patient by <15yrs, Male  \
0                                                  0              
1                                                  0              
2                                                  0              
3                                                 32              
4                                                  7              

  6 or more months of ARVs dispensed to patient by 15+yrs, Female  \
0                                                  0                
1                                                118                
2                                                  0                
3                                                  0                
4                                                406                

  6 or more months of ARVs dispensed to patient by 15+yrs, Male  \
0                                                 24              
1                                                218              
2                                                  0              
3                                                  0              
4                                                195              

   subcounty_ART_total pct_art_pop_subcounty usaid cdc  dod  
0                140.0              0.871429     0   0    1  
1               1022.0              0.677104     0   0    1  
2                536.0                   NaN     0   1    0  
3                240.0              1.000000     0   1    0  
4               1140.0              0.690351     1   0    0  

[5 rows x 80 columns]

In [28]:
import re

In [29]:
art['Subcounty'] = art['PIRS Subcounty'].str.extract(r'^(.*) Subcounty$', expand=False)
art['Subcounty'] = art['Subcounty'].fillna(art['PIRS Subcounty'])

In [30]:
art['Subcounty'].value_counts()

Subcounty
Central Division         43
Northern Division        18
Southern Division        15
Kawempe Division         14
Nakawa Division          13
                         ..
Katerera Town Council     1
Katete                    1
Kathile Town Council      1
Kuluba                    1
Zombo Town Council        1
Name: count, Length: 1313, dtype: int64

In [31]:
art['District'] = art['PIRS District '].str.replace(r'\s*District\s*$', '', regex=True).str.strip()

In [32]:
art.head()

Statistical Region UAIS 2011_ Region  PIRS UnitID     PIRS District   \
0    Northern Region        North East  W3hXYmtaQ4Q    Kotido District   
1    Northern Region      Mid Northern  auzuV39xOTU     Pader District   
2     Eastern Region        North East  J2JiemiIQvD   Katakwi District   
3     Eastern Region        North East  aVQCbe80pl7    Serere District   
4    Northern Region      Mid Northern  yDPmjwoOIxi  Alebtong District   

         PIRS Subcounty                     PIRS Service outlet  \
0  Nakapelimo Subcounty           405 Brigade Health Centre III   
1       Pader Subcounty  5th Military Division Health Centre IV   
2        Usuk Subcounty                  Aakum Health Centre II   
3      Labori Subcounty               Aarapoo Health Centre III   
4       Abako Subcounty                 Abako Health Centre III   

  PIRS Mechanism name  pirs Mechanism ID PIRS US Agency     DATIM ID  ...  \
0        URC DOD/UPDF           160698.0            DOD  OELkOLxvZxB  ...   
1        URC DOD/UPDF           160698.0            DOD  OELkOLxvZxB  ...   
2  AIC Soroti  Region            87041.0            CDC  l86BIz6pDzr  ...   
3  AIC Soroti  Region            87041.0            CDC  JXTSTKGSHxq  ...   
4   JCRC Kigezi Lango            81997.0          USAID  vitFG8P7w6d  ...   

  6 or more months of ARVs dispensed to patient by <15yrs, Male  \
0                                                  0              
1                                                  0              
2                                                  0              
3                                                 32              
4                                                  7              

  6 or more months of ARVs dispensed to patient by 15+yrs, Female  \
0                                                  0                
1                                                118                
2                                                  0                
3                                                  0                
4                                                406                

  6 or more months of ARVs dispensed to patient by 15+yrs, Male  \
0                                                 24              
1                                                218              
2                                                  0              
3                                                  0              
4                                                195              

  subcounty_ART_total pct_art_pop_subcounty  usaid cdc dod   Subcounty  \
0               140.0              0.871429      0   0   1  Nakapelimo   
1              1022.0              0.677104      0   0   1       Pader   
2               536.0                   NaN      0   1   0        Usuk   
3               240.0              1.000000      0   1   0      Labori   
4              1140.0              0.690351      1   0   0       Abako   

   District  
0    Kotido  
1     Pader  
2   Katakwi  
3    Serere  
4  Alebtong  

[5 rows x 82 columns]

In [33]:
art.shape

(2179, 82)

In [34]:
art_pop = pd.merge(art, merged_df, on=['Subcounty', 'District'], how='left')

In [35]:
art_pop

Statistical Region UAIS 2011_ Region  PIRS UnitID     PIRS District   \
0       Northern Region        North East  W3hXYmtaQ4Q    Kotido District   
1       Northern Region      Mid Northern  auzuV39xOTU     Pader District   
2        Eastern Region        North East  J2JiemiIQvD   Katakwi District   
3        Eastern Region        North East  aVQCbe80pl7    Serere District   
4       Northern Region      Mid Northern  yDPmjwoOIxi  Alebtong District   
...                 ...               ...          ...                ...   
2174     Central Region           Kampala  U6DZjzffCuQ   Kampala District   
2175     Central Region           Kampala  aB3wfoQaDTI   Kampala District   
2176     Central Region           Kampala  nVK7vHSkLbr   Kampala District   
2177     Central Region           Kampala  WuohOAlguVz   Kampala District   
2178     Central Region           Kampala  GuJvMV22ihs   Kampala District   

            PIRS Subcounty                        PIRS Service outlet  \
0     Nakapelimo Subcounty              405 Brigade Health Centre III   
1          Pader Subcounty     5th Military Division Health Centre IV   
2           Usuk Subcounty                     Aakum Health Centre II   
3         Labori Subcounty                  Aarapoo Health Centre III   
4          Abako Subcounty                    Abako Health Centre III   
...                    ...                                        ...   
2174      Kawempe Division      Mulago NRH - MARPI STI Project Clinic   
2175      Kawempe Division  Mulago NRH - Infectious Disease Institute   
2176      Kawempe Division                  Mulago NRH - MUJHU Clinic   
2177      Kawempe Division        Mulago NRH - PIDC COE Baylor Clinic   
2178      Kawempe Division             Mulago NRH - Ward 5&6(TB Ward)   

     PIRS Mechanism name  pirs Mechanism ID PIRS US Agency     DATIM ID  ...  \
0           URC DOD/UPDF           160698.0            DOD  OELkOLxvZxB  ...   
1           URC DOD/UPDF           160698.0            DOD  OELkOLxvZxB  ...   
2     AIC Soroti  Region            87041.0            CDC  l86BIz6pDzr  ...   
3     AIC Soroti  Region            87041.0            CDC  JXTSTKGSHxq  ...   
4      JCRC Kigezi Lango            81997.0          USAID  vitFG8P7w6d  ...   
...                  ...                ...            ...          ...  ...   
2174  ROM Kampala Region            87040.0            CDC  BQK9syjDN88  ...   
2175  ROM Kampala Region            87040.0            CDC  fUNiWVFY6Gi  ...   
2176  ROM Kampala Region            87040.0            CDC  hEBic79Sndd  ...   
2177  ROM Kampala Region            87040.0            CDC  jpoilGAwPPN  ...   
2178  ROM Kampala Region            87040.0            CDC  uGSDnIsoB96  ...   

     6 or more months of ARVs dispensed to patient by 15+yrs, Male  \
0                                                    24              
1                                                   218              
2                                                     0              
3                                                     0              
4                                                   195              
...                                                 ...              
2174                                                353              
2175                                               1588              
2176                                                220              
2177                                               1841              
2178                                                  0              

     subcounty_ART_total pct_art_pop_subcounty usaid cdc  dod  \
0                  140.0              0.871429     0   0    1   
1                 1022.0              0.677104     0   0    1   
2                  536.0                   NaN     0   1    0   
3                  240.0              1.000000     0   1    0   
4                 1140.0              0.690351     1   0    0   
...        

In [36]:
print(art.columns.tolist())

['Statistical Region', 'UAIS 2011_ Region', 'PIRS UnitID', 'PIRS District ', 'PIRS Subcounty', 'PIRS Service outlet', 'PIRS Mechanism name', 'pirs Mechanism ID', 'PIRS US Agency', 'DATIM ID', 'DHIS2 ID', 'DHIS2 District', 'DHIS2 Subcounty', 'DHIS2 HF Name', 'COP  Mechanism name', 'COP  Mechanism ID', 'COP US Agency', 'Type of Support', 'Period', 'Numerator (Overall TOTAL) ', 'Unknown age Female', 'Female\n<1yr', 'Female\n1-4yrs', 'Female\n5-9yrs', 'Female\n10-14 yrs', 'Female\n15-19 yrs', 'Female\n20-24 yrs', 'Female\n25-29 yrs', 'Female\n30-34 yrs', 'Female\n35-39 yrs', 'Female\n40-44 yrs', 'Female\n45-49 yrs', 'Female\n 50-54 yrs', 'Female\n 55-59 yrs', 'Female\n 60-64 yrs', 'Female\n 65+ yrs', 'Male\nUnknown age ', 'Male\n<1yr', 'Male\n1-4yrs', 'Male\n5-9yrs', 'Male\n10-14 yrs', 'Male\n15-19 yrs', 'Male\n20-24 yrs', 'Male\n25-29 yrs', 'Male\n30-34 yrs', 'Male\n35-39 yrs', 'Male\n40-44 yrs', 'Male\n45-49 yrs', 'Male\n 50-54 yrs', 'Male\n 55-59 yrs', 'Male\n 60-64 yrs', 'Male\n 65+ yr

In [37]:
art_pop.groupby('Subcounty')['Subcounty Population'].apply(lambda x: x.isna().sum()).sum()

912

In [38]:
art_pop.groupby('Subcounty')['Subcounty Population'].apply(lambda x: x.notna().sum()).sum()

1227

In [39]:
art_pop.shape

(2179, 84)

In [40]:
art_pop['Subcounty'].value_counts()

Subcounty
Central Division         43
Northern Division        18
Southern Division        15
Kawempe Division         14
Nakawa Division          13
                         ..
Katerera Town Council     1
Katete                    1
Kathile Town Council      1
Kuluba                    1
Zombo Town Council        1
Name: count, Length: 1313, dtype: int64

In [41]:
matched_rows = art_pop[art_pop['Subcounty Population'].notna()]
unmatched_rows = art_pop[art_pop['Subcounty Population'].isna()]

In [42]:
#Create a version of merged_df with Subcounties that occur only once
unique_subcounties = merged_df['Subcounty'].value_counts()
single_entry_subcounties = unique_subcounties[unique_subcounties == 1].index
merged_df_single = merged_df[merged_df['Subcounty'].isin(single_entry_subcounties)]

In [43]:
# Step 4: Try merging the unmatched rows again, this time only on 'Subcounty'
retry_merge = pd.merge(
    unmatched_rows.drop(columns=[col for col in merged_df.columns if col in unmatched_rows.columns and col != 'Subcounty']),
    merged_df_single,
    on='Subcounty',
    how='left'
)

In [44]:
# Step 5: Combine the matched rows and re-merged unmatched rows
final_merged = pd.concat([matched_rows, retry_merge], ignore_index=True)


In [45]:
final_merged.groupby('Subcounty')['Subcounty Population'].apply(lambda x: x.isna().sum()).sum()

840

In [46]:
final_merged

Statistical Region UAIS 2011_ Region  PIRS UnitID    PIRS District   \
0        Eastern Region        North East  J2JiemiIQvD  Katakwi District   
1        Eastern Region        North East  aVQCbe80pl7   Serere District   
2        Eastern Region        North East  IswqGyDbcgm   Amuria District   
3        Eastern Region        North East  M864p5iRGub   Amuria District   
4        Central Region           Kampala  aWtkpwtp9JE  Kampala District   
...                 ...               ...          ...               ...   
2174    Northern Region         West Nile  axmLtifg8dz    Yumbe District   
2175    Northern Region         West Nile  Gs4nX12jUeA    Yumbe District   
2176    Northern Region         West Nile  MUxAV5pwDbk    Yumbe District   
2177    Northern Region         West Nile  Wk8yKMCUZOk    Zombo District   
2178    Northern Region         West Nile  tuYjF238GTf    Zombo District   

           PIRS Subcounty               PIRS Service outlet  \
0          Usuk Subcounty            Aakum Health Centre II   
1        Labori Subcounty         Aarapoo Health Centre III   
2     Abarilela Subcounty       Abarilela Health Centre III   
3        Orungo Subcounty           Adakun Health Centre II   
4        Central Division          Adventist Medical Centre   
...                   ...                               ...   
2174     Kululu Subcounty            Yoyo Health Centre III   
2175   Yumbe Town Council            Yumbe Health Centre IV   
2176    Kuru Town Council  Yumbe Regional Referral Hospital   
2177        Zeu Subcounty             Zeu Health Centre III   
2178   Zombo Town Council           Zumbo Health Centre III   

       PIRS Mechanism name  pirs Mechanism ID PIRS US Agency     DATIM ID  \
0       AIC Soroti  Region            87041.0            CDC  l86BIz6pDzr   
1       AIC Soroti  Region            87041.0            CDC  JXTSTKGSHxq   
2       AIC Soroti  Region            87041.0            CDC  jj0VP3cAaCp   
3                     UPMB            84980.0            CDC  BPZNP5By5hb   
4                     UPMB            84980.0            CDC  FuinxNi7q0E   
...                    ...                ...            ...          ...   
2174         IDI West Nile            84981.0            CDC  EdQRUTWUNAd   
2175         IDI West Nile            84981.0            CDC  dEj9vXe7Hw5   
2176  MoH G2G RRH Strategy            70367.0            CDC  JReRV3g7n9K   
2177         IDI West Nile            84981.0            CDC  AmnFEwjA2t5   
2178                  UPMB            84980.0            CDC  rVHhMgI2Tju   

      ... 6 or more months of ARVs dispensed to patient by 15+yrs, Male  \
0     ...                                                  0              
1     ...                                                  0              
2     ...                                                 29              
3     ...                                                  1              
4     ...                                                 12              
...   ...                                                ...              
2174  ...                                                 12              
2175  ...                                                114              
2176  ...                                                 53              
2177  ...                                                113              
2178  ...                                                 45              

     subcounty_ART_total pct_art_pop_subcounty usaid cdc  dod  \
0                  536.0                   NaN     0   1    0   
1                  240.0              1.000000     0   1    0   
2                  162.0                   NaN     0   1    0   
3                    0.0                   NaN     0   1    0   
4                19711.0              0.008168     0   1    0   
...                  ...                   ...   ...  ..  ...   
2174               193.0              0.419689     0  

In [47]:
final_merged.shape

(2179, 84)

In [48]:
missing_rows = final_merged[final_merged['Subcounty Population'].isna()]

In [49]:
#!pip install "fuzzywuzzy[speedup]"

In [50]:
from fuzzywuzzy import process

In [66]:
def get_best_fuzzy_match(pirs_row, merged_df, threshold=97):
    subcounty = str(pirs_row['PIRS Subcounty']) if pd.notna(pirs_row['PIRS Subcounty']) else ""

    # Handle edge case where subcounty is empty
    if subcounty == "":
        return None, None

    # Get the best fuzzy match for the subcounty from the entire merged_df
    result = process.extractOne(subcounty, merged_df['Subcounty'].dropna().astype(str))

    if result:  # Check if result is not None
        best_match = result[0]
        score = result[1]
        
        if score >= threshold:
            matched_row = merged_df[merged_df['Subcounty'] == best_match].iloc[0]
            return matched_row['Subcounty Population'], best_match
    return None, None

In [67]:
missing_rows[['Estimated Population', 'Best Match Subcounty']] = missing_rows.apply(
    lambda row: pd.Series(get_best_fuzzy_match(row, merged_df)), axis=1
)

/var/folders/mb/f5zh4qyd6sbf171lrlnmff040000gn/T/ipykernel_38873/889341105.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_rows[['Estimated Population', 'Best Match Subcounty']] = missing_rows.apply(


In [68]:
pd.set_option('display.max_rows', 200)  # You can change 200 to any number you want

In [70]:
missing_rows['Best Match Subcounty'].value_counts()

Best Match Subcounty
Northern Division        15
Industrial Divison       10
Central Division          8
Southern Division         4
Eastern Division          4
Luwero Town Council       3
Kajjansi Town Council     3
Ishaka Divison            2
Rwimi Town Council        2
Kambuga Town Council      1
Kebisoni Town Council     1
Kibaale Town Council      1
Name: count, dtype: int64

In [71]:
final_merged.loc[final_merged['Subcounty Population'].isna(), 'Subcounty Population'] = missing_rows['Estimated Population']

In [72]:
final_merged['Subcounty Population'].value_counts()

Subcounty Population
24818.0     16
677330.0    14
634046.0    13
786016.0    10
40395.0     10
            ..
72352.0      1
17756.0      1
27588.0      1
13160.0      1
19211.0      1
Name: count, Length: 793, dtype: int64

In [73]:
final_merged.groupby('Subcounty')['Subcounty Population'].apply(lambda x: x.isna().sum()).sum()

786

In [74]:
final_merged.to_csv('uganda_art_with_pop.csv', index=False)

## Here, I took the data to Google Sheets and manually cleaned the 786 rows that did not merge automatically with population statistics by comparing the subcounty and district names with the subcounty population data from the 2014 census. About 200 rows still had no likely match. I brought the dataset back below.

In [75]:
df = pd.read_csv('uganda_art_with_pop_revised.csv')

In [77]:
df.head()

Statistical Region UAIS 2011_ Region  PIRS UnitID     PIRS District  \
0     Eastern Region        North East  J2JiemiIQvD  Katakwi District   
1     Eastern Region        North East  aVQCbe80pl7   Serere District   
2     Eastern Region        North East  IswqGyDbcgm   Amuria District   
3     Eastern Region        North East  M864p5iRGub   Amuria District   
4     Central Region           Kampala  aWtkpwtp9JE  Kampala District   

        PIRS Subcounty          PIRS Service outlet PIRS Mechanism name  \
0       Usuk Subcounty       Aakum Health Centre II  AIC Soroti  Region   
1     Labori Subcounty    Aarapoo Health Centre III  AIC Soroti  Region   
2  Abarilela Subcounty  Abarilela Health Centre III  AIC Soroti  Region   
3     Orungo Subcounty      Adakun Health Centre II                UPMB   
4     Central Division     Adventist Medical Centre                UPMB   

   pirs Mechanism ID PIRS US Agency     DATIM ID  ...  \
0            87041.0            CDC  l86BIz6pDzr  ...   
1            87041.0            CDC  JXTSTKGSHxq  ...   
2            87041.0            CDC  jj0VP3cAaCp  ...   
3            84980.0            CDC  BPZNP5By5hb  ...   
4            84980.0            CDC  FuinxNi7q0E  ...   

  6 or more months of ARVs dispensed to patient by 15+yrs, Male  \
0                                                  0              
1                                                  0              
2                                                 29              
3                                                  1              
4                                                 12              

  subcounty_ART_total pct_art_pop_subcounty usaid cdc  dod         Subcounty  \
0               536.0                   NaN     0   1    0              Usuk   
1               240.0              1.000000     0   1    0            Labori   
2               162.0                   NaN     0   1    0         Abarilela   
3                 0.0                   NaN     0   1    0            Orungo   
4             19711.0              0.008168     0   1    0  Central Division   

  District  County  Subcounty Population  
0  Katakwi    Usuk               21224.0  
1   Serere  Kasilo               19877.0  
2   Amuria  Amuria               24290.0  
3   Amuria  Amuria               14758.0  
4  Kampala    KCCA              150336.0  

[5 rows x 84 columns]

In [78]:
df['Subcounty Population'].isna().sum()

217

In [82]:
df['subcounty_pop_adjusted'] = df['Subcounty Population'] * 1.03

In [83]:
df.head()

Statistical Region UAIS 2011_ Region  PIRS UnitID     PIRS District  \
0     Eastern Region        North East  J2JiemiIQvD  Katakwi District   
1     Eastern Region        North East  aVQCbe80pl7   Serere District   
2     Eastern Region        North East  IswqGyDbcgm   Amuria District   
3     Eastern Region        North East  M864p5iRGub   Amuria District   
4     Central Region           Kampala  aWtkpwtp9JE  Kampala District   

        PIRS Subcounty          PIRS Service outlet PIRS Mechanism name  \
0       Usuk Subcounty       Aakum Health Centre II  AIC Soroti  Region   
1     Labori Subcounty    Aarapoo Health Centre III  AIC Soroti  Region   
2  Abarilela Subcounty  Abarilela Health Centre III  AIC Soroti  Region   
3     Orungo Subcounty      Adakun Health Centre II                UPMB   
4     Central Division     Adventist Medical Centre                UPMB   

   pirs Mechanism ID PIRS US Agency     DATIM ID  ... subcounty_ART_total  \
0            87041.0            CDC  l86BIz6pDzr  ...               536.0   
1            87041.0            CDC  JXTSTKGSHxq  ...               240.0   
2            87041.0            CDC  jj0VP3cAaCp  ...               162.0   
3            84980.0            CDC  BPZNP5By5hb  ...                 0.0   
4            84980.0            CDC  FuinxNi7q0E  ...             19711.0   

  pct_art_pop_subcounty usaid cdc dod         Subcounty District  County  \
0                   NaN     0   1   0              Usuk  Katakwi    Usuk   
1              1.000000     0   1   0            Labori   Serere  Kasilo   
2                   NaN     0   1   0         Abarilela   Amuria  Amuria   
3                   NaN     0   1   0            Orungo   Amuria  Amuria   
4              0.008168     0   1   0  Central Division  Kampala    KCCA   

  Subcounty Population  subcounty_pop_adjusted  
0              21224.0                21860.72  
1              19877.0                20473.31  
2              24290.0                25018.70  
3              14758.0                15200.74  
4             150336.0               154846.08  

[5 rows x 85 columns]

In [84]:
df.to_csv('uganda_art_with_pop_revised.csv',index=False)

In [85]:
pop = df[['PIRS District','PIRS Subcounty','subcounty_pop_adjusted']]

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


In [87]:
pop.to_csv('subcounty_population_adjusted.csv',index=False)